### PR TITLE
[#730] Skip generation of else and last case if all cases are covered.

### PR DESCRIPTION
**Pull request by ArneDeutsch**
Signed-off-by: Arne Deutsch <Arne.Deutsch@itemis.de>

### CODE REVIEW

**Comment by ArneDeutsch** [comment]
When reviewing plase have an eye on the following questions:
# Is there a risk that performance suffers? I have used some method calls in the first loop from which I can not judge well the performance costs.
# Is the missing return statement for the last case an issue (when the else is missing). Or good as it is?
# Do you have better ideas for the names "voidIncluded" and "notNullIncluded"? They are a bit technical, but I do not understand it well enough to find better names.

File: org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/compiler/DispatchCompilerTest.xtend [hunk 1]
```diff
@@ -497,5 +497,120 @@ class DispatchCompilerTest extends AbstractXtendCompilerTest {
...
+			  }
+			  
+			  protected void _doThing(final Object o) {
+			  }
+			  
+			  public void doThing(final Object v) {
+			    if (v == null) {
+			      _doThing((Void)null);
+			      return;
+			    } else _doThing(v);
```
Comment (szarnekow) [inline org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/compiler/DispatchCompilerTest.xtend:521]: Nit: Can you wrap the else statement in a block?

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java [hunk 2]
```diff
@@ -58,23 +58,53 @@ protected void initialize(JvmOperation dispatchOperation) {
...
+				JvmTypeReference caseParameterType = caseParam.getParameterType();
+				LightweightTypeReference caseParamType = owner.toLightweightTypeReference(caseParam.getParameterType());
 				if (!Strings.equal(dispatchParameterType.getIdentifier(), caseParameterType.getIdentifier())) {
 					allCasesSameType[i] = false;
-					break;
+				}
+				if (caseParamType.isType(Void.class)) {
+					voidIncluded[i] = true;
+				}
+				TypeConformanceComputationArgument rawNoSynonyms = new TypeConformanceComputationArgument(true, false, true, true, false, false);
```
Comment (szarnekow) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:86]: You might want to use `LightweightTypeReference.getIdentifier()` similar to what is done in `XbaseValidator.checkCast(JvmTypeReference, LightweightTypeReference, LightweightTypeReference)` in the last condition.
But I see that this is basically reused code from below. Can you please extract a private method for that? Something like `isSameType(..)` or `needsTypeGuard(..)`?

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java [hunk 3]
```diff
@@ -58,23 +58,53 @@ protected void initialize(JvmOperation dispatchOperation) {
...
 		boolean[] allCasesSameType = new boolean[parameterCount];
+		boolean[] voidIncluded = new boolean[parameterCount];
+		boolean[] notNullIncluded = new boolean[parameterCount];
 		for(int i = 0; i < parameterCount; i++) {
 			allCasesSameType[i] = true;
-			JvmTypeReference dispatchParameterType = dispatchOperation.getParameters().get(i).getParameterType();
+			voidIncluded[i] = false;
+			notNullIncluded[i] = false;
+			JvmFormalParameter dispatchParam = dispatchOperation.getParameters().get(i);
+			LightweightTypeReference dispatchParamType = owner.toLightweightTypeReference(dispatchParam.getParameterType());
```
Comment (szarnekow) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:74]: I find these names hard to distinguish - dispatchParamType vs dispatchParameterType

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java [hunk 4]
```diff
@@ -58,23 +58,53 @@ protected void initialize(JvmOperation dispatchOperation) {
...
+			JvmTypeReference dispatchParameterType = dispatchParam.getParameterType();
 			for (JvmOperation operation : sortedDispatchOperations) {
-				JvmFormalParameter parameter = operation.getParameters().get(i);
-				JvmTypeReference caseParameterType = parameter.getParameterType();
+				JvmFormalParameter caseParam = operation.getParameters().get(i);
+				JvmTypeReference caseParameterType = caseParam.getParameterType();
+				LightweightTypeReference caseParamType = owner.toLightweightTypeReference(caseParam.getParameterType());
 				if (!Strings.equal(dispatchParameterType.getIdentifier(), caseParameterType.getIdentifier())) {
```
Comment (szarnekow) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:80]: It should be possible to use getIdentifier from the LightweightTypeReference here and thereby avoid the need for the JvmTypeReference local vars.

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java [hunk 5]
```diff
@@ -58,23 +58,53 @@ protected void initialize(JvmOperation dispatchOperation) {
...
 					allCasesSameType[i] = false;
-					break;
+				}
+				if (caseParamType.isType(Void.class)) {
+					voidIncluded[i] = true;
```
Comment (szarnekow) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:84]: Note to self: `voidIncluded` implies that there was a null check against the parameter. `notNullIncluded` implies that that the exact type was used.

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java [hunk 6]
```diff
@@ -58,23 +58,53 @@ protected void initialize(JvmOperation dispatchOperation) {
...
+		}
+		if (parameterCount == 1 && sortedDispatchOperations.size() == 1)
+			needsElse = false;
+		for (boolean n : notNullIncluded) {
+			if (!n) {
+				needsLastCase = true;
+				break;
+			}
+		}
+		if (needsElse || (parameterCount == 1 && sortedDispatchOperations.size() == 1))
```
Comment (szarnekow) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:106]: Line 99 sets `needsElse` to false if 
`parameterCount == 1 && sortedDispatchOperations.size() == 1` - here we have the same condition - needsElse is already provable false in that case. We can probably rewrite this to:
```
if (parameterCount == 1 && sortedDispatchOperations.size() == 1) {
  needsElse = false; needsLastCase = true;
} else {
  needsElse = !allTrue(voidIncluded);
  needsLastCase = !allTrue(notNullIncluded);
}
```

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java [hunk 7]
```diff
@@ -58,23 +58,53 @@ protected void initialize(JvmOperation dispatchOperation) {
...
-		ITypeReferenceOwner owner = new StandardTypeReferenceOwner(services, dispatchOperation);
+		for (boolean v : voidIncluded) {
+			if (!v) {
+				needsElse = true;
+				break;
+			}
+		}
+		if (parameterCount == 1 && sortedDispatchOperations.size() == 1)
+			needsElse = false;
+		for (boolean n : notNullIncluded) {
```
Comment (szarnekow) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:100]: Please extract a tiny method such that we can write
`needsElse = !allTrue(voidIncluded)` and `needsLastCase = !allTrue(notNullIncluded)`.
Please try to find a better name for `allTrue(boolean[] booleans)`

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java [hunk 8]
```diff
@@ -58,23 +58,53 @@ protected void initialize(JvmOperation dispatchOperation) {
 	public void apply(/* @Nullable */ ITreeAppendable a) {
 		if (a == null)
 			throw new IllegalArgumentException("a is never null");
-		boolean needsElse = true;
+		boolean needsElse = false;
```
Comment (szarnekow) [inline org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:61]: Please move the declarations of these locals closer to their first usage

**Comment by szarnekow** [comment]
@ArneDeutsch  The approach looks fair enough to me. The comments are all only related to implementation details. I'm aware that the code was already messy as is. Nevertheless I'd greatly appreciate if you could extract a few smaller methods as outlined.

**Comment by szarnekow** [comment]
> Is the missing return statement for the last case an issue (when the else is missing). Or good as it is?

I didn't spot that immediately. Can you elaborate? Which return is missing?

File: org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/compiler/DispatchCompilerTest.xtend (hunk 1 above)
Comment (szarnekow) [inline org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/compiler/DispatchCompilerTest.xtend:521]: You mean this location where there is no return?

Comment (ArneDeutsch) [reply org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/compiler/DispatchCompilerTest.xtend:521]: Yes, that is what I mean.

Comment (szarnekow) [reply org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/compiler/DispatchCompilerTest.xtend:521]: Perfectly fine for me.

Comment (ArneDeutsch) [reply org.eclipse.xtend.core.tests/src/org/eclipse/xtend/core/tests/compiler/DispatchCompilerTest.xtend:521]: Will do.

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java (hunk 8 above)
Comment (ArneDeutsch) [reply org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:61]: Will do.

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java (hunk 3 above)
Comment (ArneDeutsch) [reply org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:74]: Me too. But to be honest ... I do not understand the code at this place at all, that's why it is hard for me to make good names here.

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java (hunk 4 above)
Comment (ArneDeutsch) [reply org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:80]: Yes, will change that. Will also solve the naming issue.

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java (hunk 2 above)
Comment (ArneDeutsch) [reply org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:86]: Will do.

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java (hunk 7 above)
Comment (ArneDeutsch) [reply org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:100]: Will do, althought a no much better name comes to mind.

File: org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java (hunk 6 above)
Comment (ArneDeutsch) [reply org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java:106]: Code does not work, leads to failing tests. But will simplify it a bit anyway.

**Review by szarnekow — APPROVED** [review]

**Pull request merged**